ppy/osu-framework
# Fix dangling audio sync threads and device-losing tests on Linux

## Dangling audio sync threads

When debugging another issue, I noticed that there would be several `AudioManager.syncAudioDevices()` threads active at a time. The reason for this, I _believe_ is to do with `AudioThreadTest` exiting the `AudioThread` before disposing the `AudioManager`, resulting in the `AudioManager`'s dispose implementation to sometimes never be reached.

## Device loss tests

Then I noticed that `DeviceLosingAudioTest` inherited from `AudioThreadTest` and hoped that fixing the above would also fix this fixture (it 100% fails on Linux), but it didn't.

### `TestPlaybackWithDeviceLoss`

`AudioManagerWithDeviceLoss.InitBass()` would keep a stale `CurrentDevice` value if the initialised device was the no-sound device. This was fixed by polling the value of `Bass.CurrentDevice` in all branches.

### `TestPlaybackWithDeviceRestore`

The general process for initialising devices goes something like this:

1. Check if the current device is the requested device and abort if so.
2. Attempt to init the device.
3. If the device is already initialised (`Errors.Already`):
    1. Select the device.
    2. **If _not_ Linux, free the device.**
    3. Re-init the device (expect anything except `Errors.Already`).
4. Return if the init succeeded (if `Bass.LastError == Errors.OK`).
5. If the init succeeded, call `UpdateDevice()` on the hierarchy.

The problem occurred on the bolded line: because the device is not freed, `Bass.LastError` will _always_ be `Errors.Already` and the method will always return `false`, in-turn bypassing step 5.  
So now there are mixers still active on the last device (because `UpdateDevice()` hasn't been called) while a new device has been selected. When the previous device is reselected, the pre-check (step 1) passes and if the device has previously been initialised then it will be freed (step 3.ii), but this also frees the mixers which were still on the device.  
The following call to change the mixer's device fails with `Errors.Handle` because the mixer has been freed internally to BASS.

This is a Linux-only issue.
